src/controller/proxy/controller.go

```diff
@@ -212,81 +230,163 @@ func manifestListContentTypeKey(rep string, art lib.ArtifactInfo) string {
+		return &manifestResult{exists, descriptor}, err
+	})
+	result := rawResult.(*manifestResult)
+	return result.exists, result.descriptor, err
 }
 
-func (c *controller) ProxyBlob(ctx context.Context, p *proModels.Project, art lib.ArtifactInfo) (int64, io.ReadCloser, error) {
+func (c *controller) ProxyBlob(ctx context.Context, _ *proModels.Project, art lib.ArtifactInfo, remote RemoteInterface) (int64, io.ReadCloser, error) {
 	remoteRepo := getRemoteRepo(art)
-	log.Debugf("The blob doesn't exist, proxy the request to the target server, url:%v", remoteRepo)
-	rHelper, err := NewRemoteHelper(ctx, p.RegistryID, WithSpeed(p.ProxyCacheSpeed()))
-	if err != nil {
-		return 0, nil, err
+
+	// If async local caching is enabled, blobs will be fetched from the remote registry and cached in the background.
+	// Otherwise, blobs will be fetched from the remove registry and cached in the local registry synchronously.
+	// Additionally, concurrent requests for the same artifact will queue and wait for the remote registry request to
+	// complete before returning data from the local registry.
+	asyncEnabled := config.EnableAsyncLocalCaching()
+	log.Debugf("AsyncLocalCaching enabled: %v, digest: %v", asyncEnabled, art.Digest)
+	if asyncEnabled {
+		return c.proxyBlobAsync(ctx, art, remote, remoteRepo)
 	}
+	return c.proxyBlobSync(ctx, art, remote, remoteRepo)
+}
+
+func (c *controller) proxyBlobAsync(_ context.Context, art lib.ArtifactInfo, remote RemoteInterface, remoteRepo string) (int64, io.ReadCloser, error) {
+	log.Debugf("The blob doesn't exist, proxy the request to the target server, url:%v", remoteRepo)
 
-	size, bReader, err := rHelper.BlobReader(remoteRepo, art.Digest)
+	size, bReader, err := remote.BlobReader(remoteRepo, art.Digest)
 	if err != nil {
 		log.Errorf("failed to pull blob, error %v", err)
 		return 0, nil, err
 	}
+
 	desc := distribution.Descriptor{Size: size, Digest: digest.Digest(art.Digest)}
 	go func() {
-		err := c.putBlobToLocal(remoteRepo, art.Repository, desc, rHelper)
+		err := c.putBlobToLocal(remoteRepo, art.Repository, desc, remote)
 		if err != nil {
 			log.Errorf("error while putting blob to local repo, %v", err)
 		}
 	}()
 	return size, bReader, nil
 }
 
-func (c *controller) putBlobToLocal(remoteRepo string, localRepo string, desc distribution.Descriptor, r RemoteInterface) error {
+func (c *controller) proxyBlobSync(ctx context.Context, art lib.ArtifactInfo, remote RemoteInterface, remoteRepo string) (int64, io.ReadCloser, error) {
+	// avoid concurrent remote requests
+	artifactKey := remoteRepo + ":" + art.Digest
+	log.Debugf("Singleflight key: %s", artifactKey)
+	_, err, _ := blobGroup.Do(artifactKey, func() (any, error) {
+		defer func() {
+			log.Debugf("Finishing Singleflight executing for key: %s", artifactKey)
+		}()
+		log.Debugf("Starting Singleflight executing for key: %s", artifactKey)
+		return nil, c.ensureBlobCached(ctx, art, remote, remoteRepo)
+	})
```

Comment:
This is primarily where the magic is occurring. Concurrent requests will wait here for the blob to be cached in the local registry, and when that's done it'll bypass the remote registry and pull from the local one.